### PR TITLE
Announcing GenStage

### PR DESCRIPTION
**Pull request by josevalim**
@fishcakez @alco @lexmag @whatyouhide @ericmj, please review it if/when you have a while.

### CODE REVIEW

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 1]
```diff
@@ -0,0 +1,358 @@
+---
+layout: post
+title: Announcing GenStage
+author: José Valim
+category: Announcements
+excerpt: GenStage is a new Elixir behaviour for exchanging events between Elixir processes with back-pressure. In this blog post we will cover the background that led us to GenStage, some example use cases, and what we are exploring for future releases. 
+---
+
+Today we are glad to announce the official release GenStage. GenStage is a new Elixir behaviour for exchanging events between Elixir processes with back-pressure. In the short-term, we expect GenStage to replace the use cases for GenEvent as well as providing a composable abstraction for consuming data from third party systems.
```
Comment (ericmj) [inline _posts/2016-07-14-announcing-genstage.markdown]: the official release *of* GenStage.

Comment (ericmj) [inline _posts/2016-07-14-announcing-genstage.markdown]: It sounds like processes have back-pressure, but it's the events really. Maybe change to "GenStage is a new Elixir behaviour for exchanging events with back-pressure between Elixir processes."?

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+
+## Background
+
+One of the original motivations for [creating and designing in Elixir was to introduce better abstractions for working with collection](https://www.youtube.com/watch?v=Lqo9-pQuRKE). Not only that, we want to provide developers interested in manipulating collections with a path to take their code from eager, to lazy, to concurrent and then distributed.
+
+Let's discuss a simple but actual example: word counting. The idea of word counting is to receive one file and count how many times each word appeards in the document. Using the `Enum` module it could be implemented as follows:
+
+```elixir
+File.read!("path/to/some/file")
+|> String.split(" ")
```
Comment (whatyouhide) [inline _posts/2016-07-14-announcing-genstage.markdown]: `String.split("\n")` to split into lines?

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+|> Enum.flat_map(fn line ->
+    String.split(line, " ")
+   end)
+|> Enum.reduce(%{}, fn word, acc ->
+    Map.update(acc, word, 1, & &1 + 1)
+   end)
+|> Enum.to_list()
+```
+
+While the solution above works fine and is efficient for small files, it is quite restrictive for large inputs as it loads the whole file into memory. If the file is a couple gigabytes, it will all be loaded into memory.
```
Comment (whatyouhide) [inline _posts/2016-07-14-announcing-genstage.markdown]: I think we can drop the "If the file is a couple gigabytes [...]" part, it doesn't feel like it's adding anything as the fact that the whole file is loaded is pretty clear from the previous sentence IMO :)

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+
+Another issue with the solution above is that the `Enum.flat_map/2` step will build a huge list, with all the words in the comment, before we effectively start counting them. Again, for a large document, this means more memory usage and a waste of processing time in building a list that will be traversed right after.
```
Comment (whatyouhide) [inline _posts/2016-07-14-announcing-genstage.markdown]: "[...] with all the words in the comment [...]" -> "[...] with all the words in the **file** [...]"?

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+
+Luckily, Elixir provides an answer to this problem, for quite some time, called streams. One of the advantage of streams is they are lazy, allowing us traverse collections item by item, in this case, line by line, instead of loading the whole data set into memory. Let's rewrite the example above to use streams:
```
Comment (whatyouhide) [inline _posts/2016-07-14-announcing-genstage.markdown]: I would rephrase this as:

"Luckily, Elixir provides a solution to this problem (and it provided it for quite some time): streams. ..."

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+  end
+
+  def handle_events(events, _from, sleeping_time) do
+    # Print events to terminal.
+    IO.inspect(events)
+
+    # Sleep the configured time.
+    Process.sleep(sleeping_time)
+
+    # We are a consumer, so we never emit items.
```
Comment (whatyouhide) [inline _posts/2016-07-14-announcing-genstage.markdown]: "we never emit <strike>items</strike> events"?

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+
+## The path forward
+
+With the release of GenStage v0.3.0, we have reached an important milestone as `GenStage` can be used as both event managers and a way to exchange events between processes, often external data sources, with back-pressure.
+
+The v0.3.0 release also includes the [`GenStage.stream`](https://hexdocs.pm/gen_stage/Experimental.GenStage.html#stream/1) function, which allows us to consume data from a GenStage as a stream, and [`GenStage.from_enumerable`](https://hexdocs.pm/gen_stage/Experimental.GenStage.html#from_enumerable/2) which allows us to use an enumerable or a stream, like `File.stream!`, as a producer. Closing the gap between stages and streams.
+
+However, we are far from done!
+
+First of all, now is the moment for the community to step in and given GenStage a try. If you have used GenEvent in the past, can it be replaced by a GenStage? Similarly, if you were planning to implement an event handling system, give GenStage a try.
```
Comment (whatyouhide) [inline _posts/2016-07-14-announcing-genstage.markdown]: `given` -> `give`

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+|> Enum.to_list()
+```
+
+For the word counting problem with a fixed data, early experiments show a linear increase in performance with a fixed overhead of 20%. In other words, a dataset that takes 60s with a single core, takes 36s on a machine with 2 cores and 18s in one with four cores. All of those gains by simply moving your computations from stream to flow. We plan to benchmark on machines with over 40 cores in the short term.
+
+If you are interested in `GenStage.Flow`, [we have written some documentation based on the prototypes we have built so far](https://hexdocs.pm/gen_stage/Experimental.GenStage.Flow.html). The code itself is coming in future GenStage releases. We will also have to consider how the `GenStage.Flow` API mirrors the functions in `Enum` and `Stream` to make the path from eager to concurrent clearer.
+
+We are very excited with the possibilities GenStage brings to developers and all new paths it allows us to explore and research. So give it a try and let us know! [GenStage, flows and more will also be the topic of my keynote at ElixirConf 2016](http://www.elixirconf.com/), so if need one more reason to attend, we will be waiting for you!
```
Comment (whatyouhide) [inline _posts/2016-07-14-announcing-genstage.markdown]: "if **you** need more reason**s** to [...]"

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+---
+
+Today we are glad to announce the official release GenStage. GenStage is a new Elixir behaviour for exchanging events between Elixir processes with back-pressure. In the short-term, we expect GenStage to replace the use cases for GenEvent as well as providing a composable abstraction for consuming data from third party systems.
+
+In this blog post we will cover the background that led us to GenStage, some example use cases, and what we are exploring for future releases. If instead you are looking for a quick reference, [check the project source code](https://github.com/elixir-lang/gen_stage) and [access its documentation](https://hexdocs.pm/gen_stage/).
+
+## Background
+
+One of the original motivations for [creating and designing in Elixir was to introduce better abstractions for working with collection](https://www.youtube.com/watch?v=Lqo9-pQuRKE). Not only that, we want to provide developers interested in manipulating collections with a path to take their code from eager, to lazy, to concurrent and then distributed.
```
Comment (ericmj) [inline _posts/2016-07-14-announcing-genstage.markdown]: ✂️ "in" in "creating and designing in Elixir was to introduce better abstractions for working with collection"

File: _posts/2016-07-14-announcing-genstage.markdown (hunk 5 above)
Comment (ericmj) [inline _posts/2016-07-14-announcing-genstage.markdown]: (and ~~it~~has provided it for quite some time)

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+```elixir
+File.read!("path/to/some/file")
+|> String.split(" ")
+|> Enum.flat_map(fn line ->
+    String.split(line, " ")
+   end)
+|> Enum.reduce(%{}, fn word, acc ->
+    Map.update(acc, word, 1, & &1 + 1)
+   end)
+|> Enum.to_list()
```
Comment (ericmj) [inline _posts/2016-07-14-announcing-genstage.markdown]: Is the `Enum.to_list` providing any value in this case? I understand that it will be needed for `Flow` later on.

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+
+  * If `Stream.async` is introducing new processes, how can we guarantee those processes are supervised?
+
+  * Since we are exchanging messages between stages, how can we do so with back-pressure?
+
+We have jumped through different abstractions trying to answer those questions until we finally settled on GenStage.
+
+## GenStage
+
+GenStage is a new Elixir behaviour for exchanging events between Elixir processes with back-pressure. Developers who use GenStage only need to worry about how the data is produced, manipulated and consumed. The act of dispatching the data and providing back-pressure is completely abstracted away from the developers.
```
Comment (ericmj) [inline _posts/2016-07-14-announcing-genstage.markdown]: Same back-pressure issue here :)

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+
+As a quick example, let's write a simple pipeline that will produce events as increasing numbers, multiply those numbers by two, and then print them them to terminal. We will do so by implementing three stages, the `:producer`, the `:producer_consumer` and the `:consumer`.
```
Comment (ericmj) [inline _posts/2016-07-14-announcing-genstage.markdown]: "them them"

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+  end
+
+  def handle_events(events, _from, number) do
+    events = Enum.map(events, & &1 * number)
+    {:noreply, events, number}
+  end
+end
+```
+
+C is the consumer which will finally receive those events and print them every second to the terminal:
```
Comment (ericmj) [inline _posts/2016-07-14-announcing-genstage.markdown]: Missing backticks around C.

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 14]
```diff
@@ -0,0 +1,358 @@
...
+      GenStage.reply(from, :ok)
+      dispatch_events(queue, demand - 1, [event | events])
+    else
+      _ -> {:noreply, Enum.reverse(events), {queue, demand}}
+    end
+  end
+end
+```
+
+The `EventManager` works as a buffer. If there is demand but not events to be sent, we store such demand. If there are events but no demand, we store such events in a queue. If a client tries to broadcast an event, the `sync_notify` call will block until the event is effectively broadcast. The bulk of the logic is in the `dispatch_events/3` function that takes events from the queue while there is demand.
```
Comment (ericmj) [inline _posts/2016-07-14-announcing-genstage.markdown]: broadcast*ed* I think.

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 15]
```diff
@@ -0,0 +1,358 @@
...
+
+By staying on control of the event manager we can configure all sorts of behaviours, including how much data we want to queue (or for how long) and if events should be buffered when there are no consumers (via the `handle_subscribe/4` and `handle_cancel/3` callbacks).
```
Comment (ericmj) [inline _posts/2016-07-14-announcing-genstage.markdown]: "By staying on control of the event manager we can configure all sorts of behaviours". I don't understand this. What does "staying on control" mean?

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 16]
```diff
@@ -0,0 +1,358 @@
...
+
+However, we are far from done!
+
+First of all, now is the moment for the community to step in and try GenStage out. If you have used GenEvent in the past, can it be replaced by a GenStage? Similarly, if you were planning to implement an event handling system, give GenStage a try.
+
+Developers who maintain libraries that integrate with external data sources, be it a RabbitMQ, Redis or Apacha Kafka, can explore GenStage as an abstraction for consuming data from those sources. Library developers must implement producers and leave it up for their users to plug the consumer stages.
+
+Once we get enough feedback, `GenStage` will be included in some shape as part of the standard library. The goal is to introduce `GenStage` and phase `GenEvent` out in the long term.
+
+We, on the Elixir team, have just get started too. The next milestone for GenStage is to revisit the original problem and provide developers a clear path to take their collection processing code from eager, to lazy, to concurrent (and then distributed).
```
Comment (ericmj) [inline _posts/2016-07-14-announcing-genstage.markdown]: have just ~~get~~got started too

**Comment by ericmj** [comment]
This is awesome @josevalim. Great article and looking forward to using GenStage.

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 17]
```diff
@@ -0,0 +1,358 @@
...
+---
+
+Today we are glad to announce the official release of GenStage. GenStage is a new Elixir behaviour for exchanging events with back-pressure between Elixir processes. In the short-term, we expect GenStage to replace the use cases for GenEvent as well as providing a composable abstraction for consuming data from third party systems.
+
+In this blog post we will cover the background that led us to GenStage, some example use cases, and what we are exploring for future releases. If instead you are looking for a quick reference, [check the project source code](https://github.com/elixir-lang/gen_stage) and [access its documentation](https://hexdocs.pm/gen_stage/).
+
+## Background
+
+One of the original motivations for [creating and designing Elixir was to introduce better abstractions for working with collection](https://www.youtube.com/watch?v=Lqo9-pQuRKE). Not only that, we want to provide developers interested in manipulating collections with a path to take their code from eager, to lazy, to concurrent and then distributed.
```
Comment (benwilson512) [inline _posts/2016-07-14-announcing-genstage.markdown]: with collection => with collections.

(lame edit sorry!)

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 18]
```diff
@@ -0,0 +1,358 @@
...
+|> Stream.async()  # NEW!
+|> Enum.reduce(%{}, fn word, acc ->
+    Map.update(acc, word, 1, & &1 + 1)
+   end)
+|> Enum.to_list()
+```
+
+The idea is that `Stream.async` would run the previous computations in a separate process that would stream its messages to the process that called `Enum.reduce`. Unfortunately, the solution above is less than ideal.
+
+First of all, we want to avoid moving data between process as much as possible. Instead, we want to start multiple processes that are performing the same computation in parallel. Not only that, if we are requiring developers to place `Stream.async` manually, it may lead to inneficient and error prone solutions.
```
Comment (benwilson512) [inline _posts/2016-07-14-announcing-genstage.markdown]: between process => between processes

Comment (benwilson512) [inline _posts/2016-07-14-announcing-genstage.markdown]: processes that are performing => processes that perform

File: _posts/2016-07-14-announcing-genstage.markdown (hunk 17 above)
Comment (alco) [inline _posts/2016-07-14-announcing-genstage.markdown]: The comma before `to lazy` is not needed.

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 19]
```diff
@@ -0,0 +1,358 @@
...
+    Map.update(acc, word, 1, & &1 + 1)
+   end)
+|> Enum.to_list()
+```
+
+While the solution above works fine and is efficient for small files, it is quite restrictive for large inputs as it loads the whole file into memory.
+
+Another issue with the solution above is that the `Enum.flat_map/2` step will build a huge list, with all the words in the file, before we effectively start counting them. Again, for a large document, this means more memory usage and a waste of processing time in building a list that will be traversed right after.
+
+Luckily, Elixir provides a solution to this problem (and has provided it for quite some time): streams. One of the advantage of streams is they are lazy, allowing us traverse collections item by item, in this case, line by line, instead of loading the whole data set into memory. Let's rewrite the example above to use streams:
```
Comment (alco) [inline _posts/2016-07-14-announcing-genstage.markdown]: `allowing us traverse` -> `allowing us to traverse`

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 20]
```diff
@@ -0,0 +1,358 @@
...
+   end)
+|> Enum.reduce(%{}, fn word, acc ->
+    Map.update(acc, word, 1, & &1 + 1)
+   end)
+|> Enum.to_list()
+```
+
+By using `File.stream!` and `Stream.flat_map`, we build a lazy computation that will emit a single line, break that line into words, and emit such words one by one without building huge lists in memory when enumerated. The functions in the [Stream module](http://elixir-lang.org/docs/stable/elixir/Stream.html) just express the computation we want to perform. The computation itself, like traversing the file or breaking into words in `flat_map`, only happen when we call a function in the `Enum` module. We have covered [the foundation for Enum and Streams](http://blog.plataformatec.com.br/2015/05/introducing-reducees/) in another article.
+
+The solution above allows us to work with large dataset without loading them all into memory. For large files, it is going to provide much better performance than the eager version. However, the solution above still does not leverage concurrency. For a machine with more than one core, which is the huge majority of machines we have available today, it is a suboptimal solution.
```
Comment (alco) [inline _posts/2016-07-14-announcing-genstage.markdown]: `large dataset` -> `large datasets`

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 21]
```diff
@@ -0,0 +1,358 @@
...
+
+First of all, we want to avoid moving data between process as much as possible. Instead, we want to start multiple processes that are performing the same computation in parallel. Not only that, if we are requiring developers to place `Stream.async` manually, it may lead to inneficient and error prone solutions.
+
+Although the solution above has many flaws, it has helped us ask the right questions:
+
+  * If `Stream.async` is introducing new processes, how can we guarantee those processes are supervised?
+
+  * Since we are exchanging messages between stages, how can we do so with back-pressure?
+
+We have jumped through different abstractions trying to answer those questions until we finally settled on GenStage.
```
Comment (alco) [inline _posts/2016-07-14-announcing-genstage.markdown]: This reads a bit weird. I would use simple past or present perfect in both parts of the sentence.

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 22]
```diff
@@ -0,0 +1,358 @@
...
+GenStage.sync_subscribe(c, to: b)
+GenStage.sync_subscribe(b, to: a)
+
+# Sleep so we see events printed.
+Process.sleep(:infinity)
+```
+
+As soon as we subscribe the stages, we should see items being printed to the terminal. Notice that, even though we have introduced a sleep command to the consumer, the producers will never overflow the consumer with data. That's because the communication between stages is demand-driven. The producer can only send items to consumers after the consumers have sent demand upstream. Producers must never send more items than consumers have asked.
+
+One of the beautiful aspects behind GenStage is that, since the communication between stages are handled by the `GenStage` behaviour itself and not the developer, parallelizing stateless stages like our consumers is really straight-forward. For example, we could do:
```
Comment (alco) [inline _posts/2016-07-14-announcing-genstage.markdown]: `the communication between stages are handled` -> `is handled`

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 23]
```diff
@@ -0,0 +1,358 @@
...
+
+During the Elixir London Meetup, I have live-coded a short example that shows how to use `GenStage` to process data stored in a PostgreSQL database.
+
+<iframe width="560" height="315" src="https://www.youtube.com/embed/aZuY5-2lwW4" frameborder="0" allowfullscreen></iframe>
+
+### GenStage for event dispatching
+
+Another scenario where GenStage can be useful today is to replace cases where developers would have used [GenEvent](http://elixir-lang.org/docs/stable/elixir/GenEvent.html) in the past. For those unfamiliar with GenEvent, it is a behaviour where events are sent to an "event manager" which then proceeds to invoke "event handlers" for each event. GenEvent, however, has one big flaw: the event manager and all events handlers run in the same process. This means GenEvent handlers cannot easily leverage concurrency without us implementing those mechanisms ourselves. Furthermore, GenEvent handlers have very ackward error semantics. Because event handlers are not separate processes, we cannot simply rely on supervisors restarting them.
```
Comment (alco) [inline _posts/2016-07-14-announcing-genstage.markdown]: `events handlers` -> `event handlers`

File: _posts/2016-07-14-announcing-genstage.markdown (hunk 14 above)
Comment (alco) [inline _posts/2016-07-14-announcing-genstage.markdown]: "broadcast" is OK too.

File: _posts/2016-07-14-announcing-genstage.markdown (hunk 15 above)
Comment (alco) [inline _posts/2016-07-14-announcing-genstage.markdown]: I've got the same question.

**Comment by josevalim** [comment]
Thanks everyone for the review. This will merged and published later today! :heart:

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 24]
```diff
@@ -0,0 +1,358 @@
...
+    end
+  end
+end
+```
+
+The `EventManager` works as a buffer. If there is demand but not events to be sent, we store such demand. If there are events but no demand, we store such events in a queue. If a client tries to broadcast an event, the `sync_notify` call will block until the event is effectively broadcast. The bulk of the logic is in the `dispatch_events/3` function that takes events from the queue while there is demand.
+
+By implementing the event manager as a producer, we can configure all sorts of behaviours that is simply not possible with `GenEvent`, such as how much data we want to queue (or for how long) and if events should be buffered or not when there are no consumers (via the `handle_subscribe/4` and `handle_cancel/3` callbacks).
+
+Implenting event handlers is as easy as implementing any other consumer. We could in fact use the `C` consumer implemented earlier. However, given event managers are often defined before the handlers, it is recommended for handlers to subscribe to managers when they start:
```
Comment (alco) [inline _posts/2016-07-14-announcing-genstage.markdown]: `Implenting` -> `Implementing`

File: _posts/2016-07-14-announcing-genstage.markdown [hunk 25]
```diff
@@ -0,0 +1,358 @@
...
+  end
+
+  def handle_events(events, _from, state) do
+    IO.inspect events
+    {:noreply, [], state}
+  end
+end
+```
+
+The guarantees that, if a supervised `EventHandler` crashes, the supervisor will start a new event handler which will promptly subscribe to the same manager, solving the ackward error handling semantics we have seen with GenEvent.
```
Comment (alco) [inline _posts/2016-07-14-announcing-genstage.markdown]: `The guarantees` -> `This guarantees`

**Comment by alco** [comment]
I've done reviewing. Great post!
